Migrate to 2018 edition.

crlf0710 opened the pull request:
- [ ] Tested on all platforms changed
- [ ] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [X] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [X] Created an example program if it would help users understand this functionality
- [X] Updated [feature matrix](https://github.com/tomaka/winit/blob/master/FEATURES.md), if new features were added or implemented

crlf0710 commented:
OK, seems the CI passed. The windows and linux migration was done by cargo-fix, and macos, android, ios, emscripten migration was done manually. Not really sure if there's anything that's covered by CI. If everything's right, this should close #912.

aleksijuvani commented:
I would imagine all this causing a lot of merge conflicts, so that's something to keep in mind. I also think we can do better here. We should be able to remove the `extern crate` declarations from `lib.rs` for example. I think we should also avoid sprinkling `crate::` everywhere and instead prefer to `use` the types in question, or use `super::` where applicable.

crlf0710 commented:
1. I have remove most of the `extern crate`s, but i'm leaving those with `macro_use`s alone.
2. I've also changed trait objects to use `dyn` syntax.
3. I've addressed the above mentioned `impl Iterator` change.

I haven't renamed the `mod.rs` yet because last time i heard that some IDE plugins doesn't support the `{module_name}.rs` project structure. Let's leave it for a subsequent PR.

felixrabe commented on src/event_loop.rs:
```diff
@@ -149,7 +149,7 @@ impl<T> EventLoop<T> {
     ///
     // Note: should be replaced with `-> impl Iterator` once stable.
```
This comment no longer applies.

crlf0710 replied:
Done

Osspial commented:
@crlf0710 Do you know which plugins, exactly? The module name change is a pretty significant ergonomics improvement for non-IDE users, and I don't think it makes sense to delay for `non-specific IDE` to support it. If we *do* end up delaying, we should at least know exactly what we're delaying for, so the change doesn't just languish in the backlog.

felixrabe commented:
@Osspial The `foo/mod.rs` -> `foo.rs` change is not mandatory. It is just a structure option added by 2018. I'm also for postponing this into an own issue to discuss this separately.

Osspial commented:
@felixrabe It's not strictly mandatory, but neither is removing `extern crate` - they're both aesthetic changes that make the code easier to read and maintain, which is why 2018 edition introduced them. I'm not necessarily opposed to delaying and discussing it, but I don't see what sort of discussion there can *be* (beyond the *this breaks the editor workflow for this IDE* issue @crlf0710 brought up) - it's clearly the direction the language is pushing, and the new structure wouldn't have been added to the language if it wasn't intended for people to adopt it.

Osspial commented on CHANGELOG.md:
```diff
@@ -296,7 +296,7 @@
 
 - Fixed windows not receiving mouse events when click-dragging the mouse outside the client area of a window, on Windows platforms.
-- Added method `os::android::EventsLoopExt:set_suspend_callback(Option<Box<Fn(bool) -> ()>>)` that allows glutin to register a callback when a suspend event happens
+- Added method `os::android::EventsLoopExt:set_suspend_callback(Option<Box<dyn Fn(bool) -> ()>>)` that allows glutin to register a callback when a suspend event happens
```
We shouldn't retroactively modify the changelog.

crlf0710 replied:
Done.

crlf0710 commented:
> Do you know which plugins, exactly?

It's either RLS or Intellj-Rust, but i can't remember which. There was also some hassle dealing with the #[path = "blah"] things.

If you wish, we can experiment again immediately in another PR after this PR is merged, and see whether there's still problems with the newest version of these plugins. I just want do these uncertain things within a separately PR so it can be easily reviewed or even reverted later if it is proved problematic.

Osspial commented:
@crlf0710 That sounds good to me.
